src/libmongoc/src/mongoc/mongoc-topology-background-monitoring.c

```diff
@@ -0,0 +1,838 @@
+      return;
+   }
+
+   /* Add newly discovered server monitors, and update existing ones. */
+   for (i = 0; i < server_descriptions->items_len; i++) {
+      mongoc_server_description_t *sd;
+
+      sd = mongoc_set_get_item (server_descriptions, i);
+      _background_monitor_reconcile_server_monitor (topology, sd);
+   }
+
+   /* Signal shutdown to server monitors no longer in the topology description.
+    */
+   server_monitor_ids_to_remove =
+      bson_malloc0 (sizeof (uint32_t) * server_monitors->items_len);
+   for (i = 0; i < server_monitors->items_len; i++) {
+      mongoc_server_monitor_t *server_monitor;
+      uint32_t id;
+
+      server_monitor = mongoc_set_get_item_and_id (server_monitors, i, &id);
+      if (!mongoc_set_get (server_descriptions, id)) {
+         if (_server_monitor_try_shutdown_and_destroy (server_monitor)) {
+            server_monitor_ids_to_remove[n_server_monitor_ids_to_remove] = id;
+            n_server_monitor_ids_to_remove++;
+         }
+      }
+   }
+
+   /* Remove freed server monitors that have completed shutdown. */
+   for (i = 0; i < n_server_monitor_ids_to_remove; i++) {
+      mongoc_set_rm (server_monitors, server_monitor_ids_to_remove[i]);
+   }
+   bson_free (server_monitor_ids_to_remove);
+}
+
+/* Request all server monitors to scan.
+ *
+ * Called from application threads (during server selection or "not master"
+ * errors). Caller must have topology mutex locked. Locks server monitor mutexes
+ * to deliver scan_requested.
+ */
+void
+_mongoc_topology_background_monitoring_request_scan (
+   mongoc_topology_t *topology)
+{
+   mongoc_set_t *server_monitors;
+   int i;
+
+   BSON_ASSERT (!topology->single_threaded);
+
+   if (topology->scanner_state == MONGOC_TOPOLOGY_SCANNER_SHUTTING_DOWN) {
+      return;
+   }
+
+   server_monitors = topology->server_monitors;
+
+   for (i = 0; i < server_monitors->items_len; i++) {
+      mongoc_server_monitor_t *server_monitor;
+      uint32_t id;
+
+      server_monitor = mongoc_set_get_item_and_id (server_monitors, i, &id);
+      _server_monitor_request_scan (server_monitor);
+   }
+}
+
+/* Stop, join, and destroy all server monitors.
+ *
+ * Called by application threads when destroying a client pool.
+ * Caller must have topology mutex locked.
+ * Locks server monitor mutexes to deliver shutdown. Releases topology mutex to
+ * join server monitor threads. Leaves topology mutex locked on exit. This
+ * function is thread-safe. But in practice, it is only ever called by one
+ * application thread (because mongoc_client_pool_destroy is not thread-safe).
+ */
+void
+_mongoc_topology_background_monitoring_stop (mongoc_topology_t *topology)
+{
+   mongoc_server_monitor_t *server_monitor;
+   int i;
+   bool is_srv_polling;
+
+   BSON_ASSERT (!topology->single_threaded);
+
+   if (topology->scanner_state == MONGOC_TOPOLOGY_SCANNER_BG_RUNNING) {
+      topology->scanner_state = MONGOC_TOPOLOGY_SCANNER_SHUTTING_DOWN;
+   } else {
+      return;
+   }
```

Comment:
Nit for readability:
```suggestion
   if (topology->scanner_state != MONGOC_TOPOLOGY_SCANNER_BG_RUNNING) {
      return;
   }

   topology->scanner_state = MONGOC_TOPOLOGY_SCANNER_SHUTTING_DOWN;
```

Reply:
Done.